src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -698,6 +698,22 @@ impl Coordinator {
                         explain_ctx,
                     })
                 } else {
+                    // If our timestamp determination relies on a timestamp, and we're not already in a
+                    // transaction, then we need to acquire read holds.
+                    if let Some(read_ts) = timestamp_context.timestamp() {
+                        if !ctx
+                            .session()
+                            .transaction()
+                            .is_in_multi_statement_transaction()
+                        {
+                            self.acquire_read_holds_auto_cleanup(
+                                ctx.session_mut(),
+                                *read_ts,
+                                &id_bundle,
+                            );
+                        }
+                    }
```

Comment:
Why not put this in `sequence_peek_timestamp` next to the other place where we get read holds? I think that would make it much easier to reason about all the scenarios where we get read holds.

Reply:
As an aside I find it confusing that we determine the timestamp and get readholds in `peek_stage_optimize_lir` and not in `peek_stage_timestamp`, but that's an existing issue.